R/author.R

```diff
@@ -0,0 +1,138 @@
+  check_is_package("use_author()")
+  maybe_name(given)
+  maybe_name(family)
+  check_character(role)
+
+  d <- proj_desc()
+  challenge_legacy_author_fields(d)
+  # We only need to consider Authors@R
+
+  authors_at_r_already <- d$has_fields("Authors@R")
+  if (authors_at_r_already) {
+    check_author_is_novel(given, family, d)
+  }
+  # This person is not already in Authors@R
+
+  author <- utils::person(given = given, family = family, role = role, ...)
+  aut_fmt <- format(author, style = 'text')
+  if (authors_at_r_already) {
+    ui_done("
+      Adding to {ui_field('Authors@R')} in DESCRIPTION:
+      {aut_fmt}")
+  } else {
+    ui_done("
+      Creating {ui_field('Authors@R')} field in DESCRIPTION and adding:
+      {aut_fmt}")
+  }
+  d$add_author(given = given, family = family, role = role, ...)
+
+  challenge_default_author(d)
+
+  d$write()
+
+  invisible(TRUE)
+
+}
+
+challenge_legacy_author_fields <- function(d = proj_desc()) {
+  has_legacy_field <- d$has_fields("Author") || d$has_fields("Maintainer")
+  if (!has_legacy_field) {
+    return(invisible())
+  }
+
+  ui_oops("
+    Found legacy {ui_field('Author')} and/or {ui_field('Maintainer')} field \\
+    in DESCRIPTION.
+    usethis only supports modification of the {ui_field('Authors@R')} field.")
+  ui_info("
+    We recommend one of these paths forward:
+    * Delete these fields and rebuild with {ui_code('use_author()')}.
+    * Convert to {ui_field('Authors@R')} with {ui_code('desc::desc_coerce_authors_at_r()')},
+      then delete the legacy fields.")
+  if (ui_yeah("Do you want to cancel this operation and sort that out first?")) {
+    ui_stop("Cancelling.")
+  }
+  invisible()
+}
+
+check_author_is_novel <- function(given = NULL, family = NULL, d = proj_desc()) {
+  authors <- d$get_authors()
+  authors_given <- purrr::map(authors, "given")
+  authors_family <- purrr::map(authors, "family")
+  m <- purrr::map2_lgl(authors_given, authors_family, function(x, y) {
+    identical(x, given) && identical(y, family)
+  })
+  if (any(m)) {
+    aut_name <- glue("{given %||% ''} {family %||% ''}")
+    usethis_abort(c(
+      "{.val {aut_name}} already appears in {.val Authors@R}.",
+      "Please make the desired change directly in DESCRIPTION or call the \\
+       desc package directly."
+    ))
+  }
+  invisible()
+}
+
+challenge_default_author <- function(d = proj_desc()) {
+  withr::with_options(
+    list(usethis.description = NULL),
+    defaults <- use_description_defaults()
+  )
+  default_author <- eval(parse(text = defaults[["Authors@R"]]))
+
+  authors <- d$get_authors()
+  m <- map_lgl(
+    authors,
+    # the `person` class is pretty weird!
+    function(x) identical(x, unclass(default_author)[[1]])
+  )
+  ui_info("
+    {ui_field('Authors@R')} appears to include a placeholder author:
+    {format(default_author, style = 'text')}")
+  if (any(m) && is_interactive() && ui_yeah("Would you like to remove it?")) {
+    # TODO: Do I want to suppress this output?
```

Comment:
I think the output is fine, I think it's nice to see confirmation that it's doing something